src/y0/algorithm/identify/idcd.py

```diff
@@ -0,0 +1,392 @@
+"""IDCD Algorithm Implementation (Algorithm 1, Lines 13-28) from Forré & Mooij (2019)."""
+
+import logging
+
+from y0.algorithm.identify import identify
+from y0.algorithm.identify.utils import (
+    Identification,
+    Unidentifiable,
+)
+
+from ..ioscm.utils import (
+    get_apt_order,
+    get_consolidated_district,
+    get_strongly_connected_components,
+)
+from ...dsl import Expression, P, Product, Variable
+from ...graph import NxMixedGraph
+
+__all__ = ["idcd"]
+
+logger = logging.getLogger(__name__)
+
+
+def idcd(
+    graph: NxMixedGraph,
+    targets: set[Variable],
+    district: set[Variable],
+    distribution: Expression,
+    *,
+    _number_recursions: int = 0,
+) -> Expression:
+    r"""Identify causal effects within consolidated districts of cyclic graphs.
+
+    This implements Algorithm 1, Lines 13-28 from [forré20a]_. IDCD is a helper function
+    called by the main ID algorithm at Line 5.
+
+    Let $G$ be a directed mixed graph (DMG), $C \subseteq D \subseteq V$ where $D$ is a
+    consolidated district with $\text{CD}(G_D) = \{D\}$, and $Q[D]$ is a probability
+    distribution over $D$. IDCD identifies the causal effect $Q[C]$ by recursively
+    shrinking the district through ancestral closure and SCC decomposition.
+
+    :param graph: The causal directed mixed graph (may contain cycles).
+    :param targets: Target set of variables to identify within district D.
+    :param district: Consolidated district containing C.
+    :param distribution: Probability distribution over district D.
+    :param _number_recursions: Recursion depth tracker for logging.
+
+    :returns: Identified causal effect Q[C].
+
+    :raises ValueError: If preconditions are violated (C ⊆ D ⊆ V).
+    :raises Unidentifiable: If causal effect cannot be identified.
+    """
+    # line 14
+    validate_preconditions(graph, targets, district, _number_recursions)
+
+    # line 15
+    subgraph_d = graph.subgraph(district)
+    ancestral_closure = compute_ancestral_closure(subgraph_d, targets, district, _number_recursions)
+
+    # # line 16
+    distribution_a = marginalize_to_ancestors(
+        subgraph_d, distribution, district, ancestral_closure, _number_recursions
+    )
+
+    # # lines 17-18
+    if ancestral_closure == targets:
+        logger.debug(f"[{_number_recursions}]: Lines 17-18 - SUCCESS (Ancestral closure = targets)")
+        return distribution_a
+
+    # # lines 19-20
+    if ancestral_closure == district:
+        logger.debug(
+            f"[{_number_recursions}]: Lines 19-20 - FAILURE (Ancestral closure = district)"
+        )
+        raise Unidentifiable(
+            f"Causal effect Q[{sorted(targets)}] is unidentifiable within district D = {sorted(ancestral_closure)}"
+        )
+
+    # checking recursive case (must have targets ⊂ ancestral_closure ⊂ district)
+
+    if not (targets < ancestral_closure < district):
```

Comment:
be really really careful that this transitive operation actually does what you think it does. There are some weird corner cases in the way python is tokenized where this might not work. would be better to just be fully explicit with a union of multiple conditionals